src/CLaSH/Class/BitPack.hs

```diff
@@ -239,6 +242,28 @@ instance (BitPack a, KnownNat (BitSize a)) => BitPack (Maybe a) where
     (c,rest) | c == low  -> Nothing
              | otherwise -> Just (unpack rest)
 
+instance (BitPack (f p)) => BitPack (M1 i c f p) where
```

Comment:
Instead of defining BitPack instances for the `GHC.Generic` datatypes, could you do the same as we do in [`CLaSH.XException`](https://github.com/clash-lang/clash-prelude/blob/3b9eadd7d6df51146cf2625043006f83f053f930/src/CLaSH/XException.hs) and define a class:

```haskell
class GBitPack f where
  type GBitSize f :: Nat
  gpack :: f a -> BitVector (GBitSize f)
  gunpack :: BitVector (GBitSize f) -> f a
```
and the instances:

```haskell
instance GBitPack U1 where ...
instance GBitPack (K1 i c) where ...
instance GBitPack (f :*: g) where ...
instance GBitPack (M m d a) where ...
```

and then give the following defaults for the `BitPack` class:

```haskell
class BitPack a where
  type BitSize a :: Nat
  type BitSize a = GBitSize (Rep a)

  pack :: a -> BitVector (BitSize a)
  default pack :: (Generic a, GBitPack (Rep a)) => a -> BitVector (GBitSize (Rep a))
  pack = gpack . from

  unpack :: BitVector (BitSize a)
  default unpack :: (Generic a, GBitPack (Rep a)) => BitVector (GBitsize (Rep a)) -> a
  unpack = to . unpack
```

That way ppl will be able to do:

```haskell
{-# DeriveGeneric, DeriveAnyClass #-}

data T where MkT ...
  deriving (Generic, BitPack)
```